File: carmajava/roadway/src/main/java/gov/dot/fhwa/saxton/carma/roadway/TransformMaintainer.java
/*
 * Copyright (C) 2017 LEIDOS.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not
 * use this file except in compliance with the License. You may obtain a copy of
 * the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations under
 * the License.
 */

package gov.dot.fhwa.saxton.carma.roadway;

import geometry_msgs.TransformStamped;
import gov.dot.fhwa.saxton.carma.geometry.GeodesicCartesianConverter;
import gov.dot.fhwa.saxton.carma.geometry.cartesian.Point3D;
import gov.dot.fhwa.saxton.carma.geometry.geodesic.Location;
import gov.dot.fhwa.saxton.carma.rosutils.SaxtonLogger;

import org.apache.commons.logging.Log;
import org.ros.message.Duration;
import org.ros.message.MessageFactory;
import org.ros.message.Time;
import org.ros.node.NodeConfiguration;
import org.ros.rosjava_geometry.Quaternion;
import org.ros.rosjava_geometry.Transform;
import org.ros.rosjava_geometry.Vector3;
import std_msgs.Header;
import tf2_msgs.TFMessage;

import java.util.Arrays;
import java.util.LinkedList;
import java.util.List;

/**
 * The TransformMaintainer if responsible for updating coordinate transforms
 */
public class TransformMaintainer {
  // Messaging and logging
  protected SaxtonLogger log;
  protected IRoadwayManager roadwayMgr;
  protected final MessageFactory messageFactory = NodeConfiguration.newPrivate().getTopicMessageFactory();

  // Host vehicle state variables
  protected boolean headingReceived = false;
  protected boolean navSatFixReceived = false;
  protected Location hostVehicleLocation = null;
  protected double hostVehicleHeading;
  // The heading of the vehicle in degrees east of north in an NED frame.
  // Frame ids
  protected final String earthFrame;
  protected final String mapFrame;
  protected final String odomFrame;
  protected final String baseLinkFrame;
  protected final String globalPositionSensorFrame;
  protected final String localPositionSensorFrame;

  // Transforms
  protected Transform mapToOdom = Transform.identity();
  protected Transform earthToMap = null;
  protected Transform baseToLocalPositionSensor = null;
  protected Transform baseToGlobalPositionSensor = null;
  protected Transform odomToBaseLink = Transform.identity();// The odom frame will start in the same orientation as the base_link frame on startup
  // Transform Update parameters
  protected Time prevMapTime = null;
  protected Duration MAP_UPDATE_PERIOD = new Duration(5); // TODO Time in seconds between updating the map frame location
  protected int tfSequenceCount = 0;

  /**
   * Constructor
   *
   * @param roadwayMgr EnvironmentWorker used to publish data and get stored transforms
   * @param log    Logging object
   * @param earthFrame The frame id used to identify the ECEF frame
   * @param mapFrame The frame id used to identify the global map frame
   * @param odomFrame The frame id used to identify the local planning frame
   * @param baseLinkFrame The frame id used to identify the host vehicle frame
   * @param globalPositionSensorFrame The frame id used to identify the frame of a global position sensor
   * @param localPositionSensorFrame The frame id used to identify a local odometry position sensor frame
   */
  public TransformMaintainer(IRoadwayManager roadwayMgr, Log log, String earthFrame, String mapFrame,
    String odomFrame, String baseLinkFrame, String globalPositionSensorFrame, String localPositionSensorFrame) {
    this.log = new SaxtonLogger(this.getClass().getSimpleName(), log);
    this.roadwayMgr = roadwayMgr;
    this.earthFrame = earthFrame;
    this.mapFrame = mapFrame;
    this.odomFrame = odomFrame;
    this.baseLinkFrame = baseLinkFrame;
    this.globalPositionSensorFrame = globalPositionSensorFrame;
    this.localPositionSensorFrame = localPositionSensorFrame;
  }

  /**
   * Handler for new vehicle heading messages
   * Headings should be specified as degrees east of north
   * TODO if base_link frame is not an +X forward frame this will need a transform as well.
   *
   * @param heading The heading message
   */
  public void handleHeadingMsg(cav_msgs.HeadingStamped heading) {
    hostVehicleHeading = heading.getHeading();
    headingReceived = true;
  }

  /**
   * NavSatFix Handler
   * Updates the host vehicle's location and updates the earth->map and map->odom transforms
   *
   * @param navSatFix The nav sat fix message
   */
  public void handleNavSatFixMsg(sensor_msgs.NavSatFix navSatFix) {
    // Assign the new host vehicle location
    hostVehicleLocation =
      new Location(navSatFix.getLatitude(), navSatFix.getLongitude(), navSatFix.getAltitude());
    navSatFixReceived = true;
    String frameId = navSatFix.getHeader().getFrameId();
    if (!frameId.equals(globalPositionSensorFrame)) {
      String msg = "NavSatFix message with unsupported frame received. Frame: " + frameId;
      log.fatal("TRANSFORM", msg);
      throw new IllegalArgumentException(msg);
    }
    updateMapAndOdomTFs();
  }

  /**
   * Helper function for use in handleNavSatFix
   * Updates the earth->map and map->odom transforms based on the current vehicle odometry, lat/lon, and heading.
   * For full functionality at least one nav sat fix and heading message needs to have been received.
   * Additionally, a transform from base_link to position_sensor needs to be available
   */
  protected void updateMapAndOdomTFs() {
    if (!navSatFixReceived || !headingReceived) {
      return; // If we don't have a heading and a nav sat fix the map->odom transform cannot be calculated
    }
    // Check if base_link->position_sensor tf is available. If not look it up
    if (baseToGlobalPositionSensor == null) {
      // This transform should be static. No need to look up more than once
      baseToGlobalPositionSensor =
        roadwayMgr.getTransform(baseLinkFrame, globalPositionSensorFrame, Time.fromMillis(0));
      if (baseToGlobalPositionSensor == null) {
        return; // If the request for this transform failed wait for another position update to request it
      }
    }

    GeodesicCartesianConverter gcc = new GeodesicCartesianConverter();

    List<geometry_msgs.TransformStamped> tfStampedMsgs = new LinkedList<>();

    // Update map location on start
    if (prevMapTime == null) { // TODO Determine earth->map update method || 0 < roadwayMgr.getTime().subtract(prevMapTime).compareTo(MAP_UPDATE_PERIOD)) {
      // Map will be an NED frame on the current vehicle location
      earthToMap = gcc.ecefToNEDFromLocaton(hostVehicleLocation);
      prevMapTime = roadwayMgr.getTime();
    }
    // Keep publishing transform to maintain timestamp
    tfStampedMsgs.add(buildTFStamped(earthToMap, earthFrame, mapFrame, roadwayMgr.getTime()));

    // Calculate map->global_position_sensor transform
    Point3D globalSensorInMap = gcc.geodesic2Cartesian(hostVehicleLocation, earthToMap.invert());

    // T_x_y = transform describing location of y with respect to x
    // m = map frame
    // b = baselink frame (from odometry)
    // B = baselink frame (from nav sat fix)
    // o = odom frame
    // p = global position sensor frame
    // We want to find T_m_o. This is the new transform from map to odom.
    // T_m_o = T_m_B * inv(T_o_b)  since b and B are both odom.
    Vector3 nTranslation = new Vector3(globalSensorInMap.getX(), globalSensorInMap.getY(), globalSensorInMap.getZ());
    // The vehicle heading is relative to NED so over short distances heading in NED = heading in map
    Vector3 zAxis = new Vector3(0, 0, 1);
    Quaternion globalSensorRotInMap = Quaternion.fromAxisAngle(zAxis, Math.toRadians(hostVehicleHeading));
    globalSensorRotInMap = globalSensorRotInMap.normalize();

    Transform T_m_p = new Transform(nTranslation, globalSensorRotInMap);
    Transform T_B_p = baseToGlobalPositionSensor;
    Transform T_m_B = T_m_p.multiply(T_B_p.invert());
    Transform T_o_b = odomToBaseLink;

    // Modify map to odom with the difference from the expected and real sensor positions
    mapToOdom = T_m_B.multiply(T_o_b.invert());
    // Publish newly calculated transforms
    tfStampedMsgs.add(buildTFStamped(mapToOdom, mapFrame, odomFrame, roadwayMgr.getTime()));
    publishTF(tfStampedMsgs);
  }

  /**
   * Odometry message handler
   *
   * @param odometry Odometry message
   */
  public void handleOdometryMsg(nav_msgs.Odometry odometry) {
    // Check if base_link->position_sensor tf is available. If not look it up
    if (baseToLocalPositionSensor == null) {
      // This transform should be static. No need to look up more than once
      baseToLocalPositionSensor =
        roadwayMgr.getTransform(baseLinkFrame, localPositionSensorFrame, Time.fromMillis(0));
      if (baseToLocalPositionSensor == null) {
        return; // If the request for this transform failed wait for another odometry update to request it
      }
    }

    String parentFrameId = odometry.getHeader().getFrameId();
    String childFrameId = odometry.getChildFrameId();
    // If the odometry is already in the base_link frame
    if (parentFrameId.equals(odomFrame) && childFrameId.equals(baseLinkFrame)) {
      odomToBaseLink = Transform.fromPoseMessage(odometry.getPose().getPose());
      publishTF(Arrays.asList(buildTFStamped(odomToBaseLink, odomFrame, baseLinkFrame, roadwayMgr.getTime())));

    } else if (parentFrameId.equals(odomFrame) && childFrameId.equals(localPositionSensorFrame)) {
      // Extract the location of the position sensor relative to the odom frame
      // Covariance is ignored as filtering was already done by sensor fusion
      // Calculate odom->base_link
      // T_x_y = transform describing location of y with respect to x
      // p = position sensor frame (from odometry)
      // o = odom frame
      // b = baselink frame (as has been calculated by odometry up to this point)
      // T_o_b = T_o_p * inv(T_b_p)
      Transform T_o_p = Transform.fromPoseMessage(odometry.getPose().getPose());
      Transform T_b_p = baseToLocalPositionSensor;
      Transform T_o_b = T_o_p.multiply(T_b_p.invert());
      odomToBaseLink = T_o_b;
      // Publish updated transform
      publishTF(Arrays.asList(buildTFStamped(odomToBaseLink, odomFrame, baseLinkFrame, roadwayMgr.getTime())));

    } else {
      String msg =
        "Odometry message with unsupported frames received. ParentFrame: " + parentFrameId
          + " ChildFrame: " + childFrameId;
      log.fatal("TRANSFORM", msg);
      throw new IllegalArgumentException(msg);
    }
  }

  /**
   * Velocity message handler TODO decide if this is needed
   *
   * @param velocity host vehicle velocity
   */
  public void handleVelocityMsg(geometry_msgs.TwistStamped velocity) {
    //TODO update host vehicle specification
  }

  /**
   * Helper function builds a tf2 message for the given transform between parent and child frames
   *
   * @param tf          The transform to publish. Describes the position of the child frame in the parent frame
   * @param parentFrame The name of the parent frame
   * @param childFrame  The name of the child frame
   * @param stamp The timestamp of this transform
   */
  protected geometry_msgs.TransformStamped buildTFStamped(Transform tf, String parentFrame,
    String childFrame, Time stamp) {
    geometry_msgs.TransformStamped tfStampedMsg =
      messageFactory.newFromType(geometry_msgs.TransformStamped._TYPE);
    Header hdr = tfStampedMsg.getHeader();
    hdr.setFrameId(parentFrame);
    hdr.setStamp(roadwayMgr.getTime());
    hdr.setSeq(tfSequenceCount);
    hdr.setStamp(stamp);
    tfStampedMsg.setChildFrameId(childFrame);
    tfStampedMsg.setTransform(tf.toTransformMessage(tfStampedMsg.getTransform()));
    return tfStampedMsg;
  }

  /**
   * Publishes a list of transforms in a single tf2 TFMessage
   *
   * @param tfList List of transforms
   */
  protected void publishTF(List<TransformStamped> tfList) {
    TFMessage tfMsg = messageFactory.newFromType(TFMessage._TYPE);
    tfMsg.setTransforms(tfList);
    tfSequenceCount++;
    roadwayMgr.publishTF(tfMsg);
  }
}

File: sensor_fusion/src/transform_maintainer.cpp
/*
 * Copyright (C) 2018-2019 LEIDOS.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not
 * use this file except in compliance with the License. You may obtain a copy of
 * the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations under
 * the License.
 */

#include "transform_maintainer.h"

namespace tf2
{
    // Functions based off tf2/buffer_core.cpp
    void transformTF2ToMsg(const tf2::Transform& tf2, geometry_msgs::Transform& msg)
    {
        msg.translation.x = tf2.getOrigin().x();
        msg.translation.y = tf2.getOrigin().y();
        msg.translation.z = tf2.getOrigin().z();
        msg.rotation.x = tf2.getRotation().x();
        msg.rotation.y = tf2.getRotation().y();
        msg.rotation.z = tf2.getRotation().z();
        msg.rotation.w = tf2.getRotation().w();
    }

    void transformTF2ToMsg(const tf2::Transform& tf2, geometry_msgs::TransformStamped& msg, ros::Time stamp, const std::string& frame_id, const std::string& child_frame_id)
    {
        transformTF2ToMsg(tf2, msg.transform);
        msg.header.stamp = stamp;
        msg.header.frame_id = frame_id;
        msg.child_frame_id = child_frame_id;
    }
}

void TransformMaintainer::nav_sat_fix_update_cb(const sensor_msgs::NavSatFixConstPtr host_veh_loc, const cav_msgs::HeadingStampedConstPtr heading_msg)
{
    
    // These values must be time synched
    if (host_veh_loc->header.stamp != heading_msg->header.stamp) {
      ROS_WARN_STREAM("TRANSFORM | Tried to update transforms without synched heading and nav_sat_fix");
      return;
    }

    if (last_nav_sat_stamp >= host_veh_loc->header.stamp) {
      ROS_WARN_STREAM("TRANSFORM | Old nav sat received prev: " << last_nav_sat_stamp << " update: " << host_veh_loc->header.stamp);
      return;
    }

    if (last_heading_stamp >= heading_msg->header.stamp) {
      ROS_WARN_STREAM("TRANSFORM | Old heading received prev: " << last_heading_stamp << " update: " << heading_msg->header.stamp);
      return;
    }

    last_nav_sat_stamp = host_veh_loc->header.stamp;
    ROS_DEBUG_STREAM("TRANSFORM | New nav sat received: " << last_nav_sat_stamp);

    last_heading_stamp = heading_msg->header.stamp;
    ROS_DEBUG_STREAM("TRANSFORM | New heading received: " << last_heading_stamp);

    std::string frame_id = host_veh_loc->header.frame_id;
    if (frame_id != global_pos_sensor_frame_) {
      std::string msg = "NavSatFix message with unsupported frame received. Frame: " + frame_id;
      ROS_ERROR_STREAM("TRANSFORM | " << msg);
      return;
    }
    
    // Check if base_link->position_sensor tf is available. If not look it up
    if (no_base_to_global_pos_sensor_) {
      // This transform should be static. No need to look up more than once
      try {
        base_to_global_pos_sensor_ = get_transform(base_link_frame_, global_pos_sensor_frame_, ros::Time(0), true);
      } catch (tf2::TransformException e) {
        std::string msg = "TransformMaintainer nav_sat_fix_update_cb failed to get transform for global position sensor in base link";
        ROS_WARN_STREAM("TRANSFORM | " << msg);
        return;
      }

      no_base_to_global_pos_sensor_ = false;
    }

    std::vector<geometry_msgs::TransformStamped> tf_stamped_msgs;

    // Extract geodesic data and convert to radians
    wgs84_utils::wgs84_coordinate host_veh_coord;
    host_veh_coord.lat = host_veh_loc->latitude * wgs84_utils::DEG2RAD;
    host_veh_coord.lon = host_veh_loc->longitude * wgs84_utils::DEG2RAD;
    host_veh_coord.elevation = host_veh_loc->altitude;
    host_veh_coord.heading = heading_msg->heading * wgs84_utils::DEG2RAD;

    // Update map location on start
    if (no_earth_to_map_) { 
      // Map will be an NED frame on the current vehicle location
      earth_to_map_ = wgs84_utils::ecef_to_ned_from_loc(host_veh_coord);
      no_earth_to_map_ = false;
    }
    // Keep publishing transform to maintain timestamp
    geometry_msgs::TransformStamped earth_to_map_msg;
    tf2::transformTF2ToMsg(earth_to_map_, earth_to_map_msg, host_veh_loc->header.stamp, earth_frame_, map_frame_);
    
    tf_stamped_msgs.push_back(earth_to_map_msg);

    // Get odom->base_link transform with matching timestamp
    tf2::Transform odom_to_base_link;
    try {
      odom_to_base_link =  get_transform(odom_frame_, base_link_frame_, host_veh_loc->header.stamp, true);
    } catch (tf2::TransformException e) {
      std::string msg = "TransformMaintainer nav_sat_fix_update_cb failed to get transform for odom to base_link";
      ROS_WARN_STREAM("TRANSFORM | " << msg);
      return;
    }

    // Calculate updated tf
    map_to_odom_ = calculate_map_to_odom_tf(
      host_veh_coord, base_to_global_pos_sensor_,
      earth_to_map_, odom_to_base_link);
    //TODO do we need to sync the heading too? Currently pinpoint's heading and nav_sat_fix are always synched. But this maynot always be the case
    // Publish newly calculated transforms
    geometry_msgs::TransformStamped map_to_odom_msg;
    tf2::transformTF2ToMsg(map_to_odom_, map_to_odom_msg, host_veh_loc->header.stamp, map_frame_, odom_frame_);
    
    tf_stamped_msgs.push_back(map_to_odom_msg);

    // Publish transform
    for (int i =0; i < tf_stamped_msgs.size(); i++) {
      tf2_buffer_->setTransform(tf_stamped_msgs.at(i), "/carma/sensor_fusion");
    }
    tf2_broadcaster_->sendTransform(tf_stamped_msgs);
}

// Broken out for unit testing
// Heading is assumed to be in rad
// All geodesic angles assumed to be in rad
  tf2::Transform TransformMaintainer::calculate_map_to_odom_tf(
    const wgs84_utils::wgs84_coordinate& host_veh_coord,
    const tf2::Transform&  base_to_global_pos_sensor, const tf2::Transform& earth_to_map,
    const tf2::Transform& odom_to_base_link)
  {
    // Calculate map->global_position_sensor transform

    tf2::Vector3 global_sensor_in_map = wgs84_utils::geodesic_to_ecef(host_veh_coord, earth_to_map.inverse());

    // T_x_y = transform describing location of y with respect to x
    // m = map frame
    // b = baselink frame (from odometry)
    // B = baselink frame (from nav sat fix)
    // o = odom frame
    // p = global position sensor frame
    // We want to find T_m_o. This is the new transform from map to odom.
    // T_m_o = T_m_B * inv(T_o_b)  since b and B are both odom.
    tf2::Vector3 sensor_trans_in_map = global_sensor_in_map;
    // The vehicle heading is relative to NED so over short distances heading in NED = heading in map
    tf2::Vector3 zAxis = tf2::Vector3(0, 0, 1);
    tf2::Quaternion sensor_rot_in_map(zAxis, host_veh_coord.heading);
    sensor_rot_in_map = sensor_rot_in_map.normalize();

    tf2::Transform T_m_p = tf2::Transform(sensor_rot_in_map, sensor_trans_in_map);
    tf2::Transform T_B_p = base_to_global_pos_sensor;
    tf2::Transform T_m_B = T_m_p * T_B_p.inverse();
    tf2::Transform T_o_b = odom_to_base_link;

    // Modify map to odom with the difference from the expected and real sensor positions
    return T_m_B * T_o_b.inverse();
  }

void TransformMaintainer::odometry_update_cb(const nav_msgs::OdometryConstPtr odometry) 
{
      // Check if base_link->position_sensor tf is available. If not look it up
    if (no_base_to_local_pos_sensor_) {
      // This transform should be static. No need to look up more than once
      try {
        base_to_local_pos_sensor_ = get_transform(base_link_frame_, local_pos_sensor_frame_, ros::Time(0), true);
      } catch (tf2::TransformException e) {
        std::string msg = "TransformMaintainer odometry_update_cb failed to get transform for local position sensor in base link";
        ROS_WARN_STREAM("TRANSFORM | " << msg);
        return;
      }

      no_base_to_local_pos_sensor_ = false;
    }

    //nav_msgs::OdometryConstPtr odometry = odom_map_->begin()->second;
    std::string parent_frame_id = odometry->header.frame_id;
    std::string child_frame_id = odometry->child_frame_id;
    // If the odometry is already in the base_link frame
    if (parent_frame_id == odom_frame_ && child_frame_id == base_link_frame_) {
      tf2::fromMsg(odometry->pose.pose, odom_to_base_link_);
      // Publish updated transform
      geometry_msgs::TransformStamped odom_to_base_link_msg; 
      tf2::transformTF2ToMsg(odom_to_base_link_, odom_to_base_link_msg, odometry->header.stamp, odom_frame_, base_link_frame_);
      tf2_buffer_->setTransform(odom_to_base_link_msg, "/carma/sensor_fusion");
      tf2_broadcaster_->sendTransform(odom_to_base_link_msg);

    } else if (parent_frame_id == odom_frame_ && child_frame_id == local_pos_sensor_frame_) {
      // Extract the location of the position sensor relative to the odom frame
      // Covariance is ignored as filtering was already done by sensor fusion
      // Calculate odom->base_link
      // T_x_y = transform describing location of y with respect to x
      // p = position sensor frame (from odometry)
      // o = odom frame
      // b = baselink frame (as has been calculated by odometry up to this point)
      // T_o_b = T_o_p * inv(T_b_p)
      tf2::Transform T_o_p;
      tf2::fromMsg(odometry->pose.pose, T_o_p);

      tf2::Transform T_b_p = base_to_local_pos_sensor_;
      tf2::Transform T_o_b = T_o_p * T_b_p.inverse();
      odom_to_base_link_ = T_o_b;
      // Publish updated transform
      geometry_msgs::TransformStamped odom_to_base_link_msg;
      tf2::transformTF2ToMsg(odom_to_base_link_, odom_to_base_link_msg, odometry->header.stamp, odom_frame_, base_link_frame_);
      tf2_buffer_->setTransform(odom_to_base_link_msg, "/carma/sensor_fusion");
      tf2_broadcaster_->sendTransform(odom_to_base_link_msg);

    } else {
      std::string msg =
        "Odometry message with unsupported frames received. ParentFrame: " + parent_frame_id
          + " ChildFrame: " + child_frame_id;
      ROS_ERROR_STREAM("TRANSFORM | " << msg);//TODO should we throw an exception here?
    }
}

// Helper function
tf2::Stamped<tf2::Transform> TransformMaintainer::get_transform(
  std::string parent_frame, std::string child_frame, ros::Time stamp,
  bool can_use_most_recent_tf) 
{
  geometry_msgs::TransformStamped transform_stamped;

  if(tf2_buffer_->canTransform(parent_frame, child_frame, stamp))
  {
    transform_stamped = tf2_buffer_->lookupTransform(parent_frame, child_frame, stamp);
  }
  else if(tf2_buffer_->canTransform(parent_frame, child_frame, ros::Time(0)) && can_use_most_recent_tf)
  {
    ROS_DEBUG_STREAM("TRANSFORM | Using latest transform available"); // This expected on the nav_sat update
    transform_stamped = tf2_buffer_->lookupTransform(parent_frame, child_frame, ros::Time(0));
  }
  else
  {
    ROS_WARN_STREAM("No transform available from " << parent_frame << " to " << child_frame);

    throw tf2::TransformException("Failed to get transform");
  }
  
  tf2::Transform result;
  tf2::fromMsg(transform_stamped.transform, result);
  tf2::Stamped<tf2::Transform> stamped_result(result, transform_stamped.header.stamp, transform_stamped.header.frame_id);
  return stamped_result;
}



